src/main/scala/com/gu/contentapi/firehose/kinesis/EventProcessor.scala

```diff
@@ -0,0 +1,119 @@
+package com.gu.contentapi.firehose.kinesis
+
+import com.amazonaws.services.kinesis.model.Record
+import com.amazonaws.services.kinesis.clientlibrary.interfaces.{ IRecordProcessor, IRecordProcessorCheckpointer }
+import com.amazonaws.services.kinesis.clientlibrary.types.ShutdownReason
+import com.typesafe.scalalogging.LazyLogging
+import com.twitter.scrooge.ThriftStruct
+import java.util.concurrent.atomic.{ AtomicInteger, AtomicLong }
+import java.util.{ List => JList }
+
+import com.gu.contentapi.firehose.kinesis.EventProcessor.EventWithSize
+
+import scala.collection.JavaConverters._
+import scala.concurrent.duration._
+import scala.util.{ Failure, Success }
+
+abstract class EventProcessor[EventT <: ThriftStruct](
+  checkpointInterval: Duration = 30.seconds,
+  maxCheckpointBatchSize: Int = 20
+)
+    extends IRecordProcessor
+    with ThriftDeserializer[EventT]
+    with LazyLogging {
+
+  private[this] var shardId: String = _
+
+  /* Use atomic to prevent any concurrent access issues */
+  private[this] val lastCheckpointedAt = new AtomicLong(System.nanoTime())
+  private[this] val recordsProcessedSinceCheckpoint = new AtomicInteger()
+
+  override def initialize(shardId: String): Unit = {
+    this.shardId = shardId
+    logger.info(s"Initialized an event processor for shard $shardId")
+  }
+
+  override def processRecords(records: JList[Record], checkpointer: IRecordProcessorCheckpointer): Unit = {
+    processRecordsIfActivated(records, checkpointer)
+  }
+
+  private def processRecordsIfActivated(records: JList[Record], checkpointer: IRecordProcessorCheckpointer): Unit = {
+    val events = records.asScala.flatMap { record =>
+      val buffer = record.getData
+      val op = deserializeEvent(buffer)
+      op match {
+        case Success(event) => Some(EventWithSize(event, buffer.array.length))
+        case Failure(e) => {
+          logger.error(s"deserialization of event buffer failed: ${e.getMessage}", e)
+          None
+        }
+      }
+    }
+
+    processEvents(events)
+
+    /*
+     Several errors can be encountered:
```

Comment:
This comment is inaccurate. It talks about compression, which is irrelevant here, and about Porter and its internals.